### ISSUE TITLE
Floating point exception when computing gradients

### ISSUE THREAD

**Issue by altaetran** [tensorflow/tensorflow#2163]
Hi tensorflow developer team! 

I just wanted to first let you know that I really appreciate all the work you are putting into this amazing open source system. It has been a terrific system for me so far. However, I have come into a roadblock when using the tf.gather function in gradients.

I am using

### Environment info
Operating System: CentOS

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
CUDA version 7.5
cuDNN version 7.0 (64 bit)
tensorflow/0.8.0-gpu version

### Steps to reproduce
I obtain a floating point exception when I run the following code

`A_ph = tf.placeholder(tf.float32, shape=(None, 2))`
`ind_ph = tf.placeholder(tf.int32, shape=(None, 2))`

`gather = tf.gather(A_ph, ind_ph)`
`redsum = tf.reduce_sum(gather, 1)`
`l2 = tf.reduce_sum(redsum)`

`A = np.array([[0,0],[0,0],[0,0]])`
`ind = np.zeros((0,2))`

`grad_op = tf.gradients(l2, A_ph)`

`out = sess.run(grad_op, feed_dict={A_ph:A, ind_ph:ind})`

### What have you tried?
If you replace ind with a non-empty array, such as [[0,1]] it works fine. 

I suspect that the output of redsum = [], which makes l2 disconnected from A_ph for the given ind. Fed forward to computing l2, I get the correct value of 0.0 (since there is nothing to be summed). However, for backpropagation of gradients, I think the output of redsum=[] chokes the algorithm, and causes it to evaluate some illegal operation using an empty tensor, instead of just backpropagating 0.0. 

Are there any solutions to this problem currently? In my use of this computation, I won't know ahead of time whether or not the intermediate computations produce [] (empty tensor), but I also want to be able to get the correct gradients when this happens. Thank you so much for your consideration!

**Comment by poxvoculi** [comment]
Thanks for the bug report.

@girving ReshapeOp is generating SIGFPE in this program.

**Comment by girving** [comment]
Thanks for reporting!  We're going to fix TensorFlow to produce nice exceptions for integer zero divisions.

**Comment by girving** [comment]
Ah, there are two bugs here: one bug that we crash the process on integer division by zero, and another that we generate a division by zero.

**Comment by altaetran** [comment]
Yes, I think there should never be any division by zero in the above operations, at least from what I can tell from the backpropagation formulas. I think I should definitely be able to get a None gradient to indicate that that l2 and A_ph are disconnected, rather than an exception. Is the bug a complex one to fix?

**Comment by girving** [comment]
@altaetran: It's conceptually straightforward: one has to go through and sanitize all the occurrences of `//` in `math_grad.py`.  Unfortunately there are a bunch of them for a variety of different ops, and the fixes look different in the different cases.

However, there's no way to get a `None` gradient, since the variables *are* connected, just through an empty tensor bottleneck.  Since we don't necessarily know that the shapes are empty until runtime, we can't produce `None`.

**Comment by girving** [comment]
@altaetran: To make my initial comment make sense: we want to produce nice error messages even for malformed ops, so we want to fix the underlying divisions by zero regardless of whether we fix the gradient code.

**Comment by altaetran** [comment]
Oh I see, so those would be two separate tasks. Is there a reason for why an empty bottle neck tensor in the gradients produces division by zero though? Or is that something to be changed as well?

**Comment by girving** [comment]
As I said: the fixes are straightforward but there are a number of them.  Here is an example of a bad line: https://github.com/tensorflow/tensorflow/blob/5e22e3a3187e6729488f4a6da59b4cfbedc40946/tensorflow/python/ops/math_grad.py#L36

**Comment by altaetran** [comment]
Oh I see ok. So then this is likely something to occur in a future patch?

**Comment by girving** [comment]
I'm looking at it now, so not too long except for the upcoming weekend. :)

I'll probably fix the gradients first since making integer division safe is slightly more awkward (to do in a reasonably performant manner).

**Comment by girving** [comment]
CL submitted, so it should be on Github within 24 hours.

**Comment by girving** [comment]
To clarify: I fixed the gradients not to divide by zero, which should solve your issue.  Integer divide by zero on its own will still crash the process.

**Comment by rbharath** [comment]
Great! Been running into a similar problem myself. As a question, is there a local version of the patch I could use? I'm running Tensorflow 0.8.0 on a cluster, and the sysadmins won't rebuild the cluster version before TF 0.9.0 (or the next stable version)

**Comment by girving** [comment]
@rbharath: Do you want source or binaries?  The source version should have the fix tomorrow.  Not sure when the nightly appears (tomorrow or the next day), but the links for those are here: https://github.com/tensorflow/tensorflow/blob/master/README.md

**Comment by rbharath** [comment]
@girving Sorry, should have been clearer. Is the fix python only? If so, would I be able to just use the corrected python module instead of the standard module without rebuilding the rest of the source? (The issue is that our cluster runs with an old CentOs with an old glibc, so the sysadmins have to do a ton of work to get tensorflow to build, so it's infeasible for us to use binaries or rebuild except at stable releases).

**Comment by girving** [comment]
@rbharath: Ah, yes: the fix I just made is Python only.  However, we don't have any setup for using different Python alongside C++, and in any case you'd have to carefully cherry-pick just my change ontop of 0.8.0 to have a hope of succeeded.
